jena-text/src/main/java/org/apache/jena/query/text/TextDatasetFactory.java

```diff
@@ -191,5 +213,50 @@ public static DatasetGraph createSolrIndex(DatasetGraph base, SolrServer server,
         TextIndex index = createSolrIndex(server, entMap) ;
         return create(base, index, true) ; 
     }
+ 
+    protected static class LuceneIndexMap {
+    	
+    	private static HashMap<Object, WeakReference<TextIndex>> map = new HashMap<Object, WeakReference<TextIndex>>();
+
+    	protected synchronized TextIndex get(Directory directory) {
```

Comment:
Thanks for the sketch.  I've tried something based on it and does look better.  Will push shortly.